domain/credential/service/service.go

```diff
@@ -124,3 +197,189 @@ func (s *Service) WatchCredential(ctx context.Context, tag names.CloudCredential
+func (s *Service) CheckAndUpdateCredential(ctx context.Context, id credential.ID, cred cloud.Credential, force bool) ([]UpdateCredentialModelResult, error) {
+	if s.validationContextGetter == nil {
+		return nil, errors.New("cannot validate credential with nil context getter")
+	}
+
+	models, err := s.credentialModels(ctx, id)
+	if err != nil && !force {
+		return nil, errors.Trace(err)
+	}
+
+	modelCloud, err := s.st.GetCloud(ctx, id.Cloud)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		modelsErred  bool
+		modelsResult []UpdateCredentialModelResult
+	)
+	if len(models) > 0 {
+		for uuid, name := range models {
+			result := UpdateCredentialModelResult{
+				ModelUUID: uuid,
+				ModelName: name,
+			}
+			result.Errors = s.validateCredentialForModel(uuid, modelCloud, id, &cred)
+			modelsResult = append(modelsResult, result)
+			if len(result.Errors) > 0 {
+				modelsErred = true
+			}
+		}
+		// Since we get a map above, for consistency ensure that models are added
+		// sorted by model uuid.
+		sort.Slice(modelsResult, func(i, j int) bool {
+			return modelsResult[i].ModelUUID < modelsResult[j].ModelUUID
+		})
+	}
+
+	if modelsErred && !force {
+		return modelsResult, credentialerrors.CredentialValidation
+	}
+
+	existingInvalid, err := s.st.UpsertCloudCredential(ctx, id.Name, id.Cloud, id.Owner, cred)
+	if err != nil {
+		if errors.Is(err, errors.NotFound) {
+			err = errors.Errorf(
+				"cannot update credential %q: controller does not manage cloud %q",
+				id.Name, id.Cloud)
+		}
+		return nil, errors.Trace(err)
+	}
+	if s.legacyUpdater == nil {
+		return modelsResult, nil
+	}
+
+	// TODO(wallyworld) - we still manage models in mongo.
+	// This can be removed after models are in dqlite.
+	// Existing credential will become valid after this call, and
+	// the model status of all models that use it will be reverted.
+	if existingInvalid != nil && *existingInvalid != cred.Invalid && !cred.Invalid {
+		tag, err := id.Tag()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if err := s.legacyUpdater(tag); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return modelsResult, nil
+}
+
+// CheckAndRevokeCredential removes the credential after first checking that any models which use the credential
+// can still access the cloud resources. If force is true, update the credential even if there are issues
+// validating the credential.
+// TODO(wallyworld) - we need a strategy to handle changes which occur after the affected models have been read
+// but before validation can complete.
+func (s *Service) CheckAndRevokeCredential(ctx context.Context, id credential.ID, force bool) error {
+	models, err := s.credentialModels(ctx, id)
+	if err != nil {
+		if !force {
+			// Could not determine if credential has models - do not continue revoking this credential...
+			return errors.Trace(err)
+		}
+		s.logger.Warningf("could not get models that use credential %v: %v", id, err)
```

Comment:
I don't think `force` should be ignoring database errors. If you don't have all the information you need that is still an error.

`force` is there to get around business logic not hard errors.

Reply:
Existing logic copied from apiserver layer. I *think* the intent was to allow a credential update to be rammed through if needed. But yeah, seems wrong to ignore db errors.